File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        
        int count = 1 << nums.length;
        
        for (int i = 0; i < count; i++) {
            List<Integer> vec = new ArrayList<>();
            for (int j = 0; j < nums.length; j++) {
                if ((i & (1 << j)) != 0) {
                    vec.add(nums[j]);
                }
            }
            res.add(new ArrayList<>(vec));
        }
        
        return res;
    }
}

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void f(int i, vector<int>& nums, vector<int>& ds, vector<vector<int>>& res, int n) {
        if (i == n) {
            res.push_back(ds);
            return;
        }
        
        ds.push_back(nums[i]);
        f(i + 1, nums, ds, res, n);
        ds.pop_back();
        f(i + 1, nums, ds, res, n);
    }
    
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        
        vector<int> ds;
        
        f(0, nums, ds, res, nums.size());
        
        return res;
    }
};